zcash_client_sqlite/src/testing/pool.rs

```diff
@@ -1599,6 +1613,212 @@ pub(crate) fn fully_funded_fully_private<P0: ShieldedPoolTester, P1: ShieldedPoo
+        .unwrap();
+    st.generate_next_block_including(*res.first());
+
+    // Generate a few more empty blocks
+    for _ in 0..5 {
+        st.generate_empty_block();
+    }
+
+    // Generate another block with funds for us
+    let (max_height, _, _) =
+        st.generate_next_block(&p0_fvk, AddressType::DefaultExternal, note_value);
+
+    // Scan everything.
+    st.scan_cached_blocks(
+        next_to_scan,
+        usize::try_from(u32::from(max_height) - u32::from(next_to_scan) + 1).unwrap(),
+    );
+
+    let expected_final =
+        (mid_balance + note_value - (transfer_amount * 3).unwrap() - (expected_fee * 3).unwrap())
+            .unwrap();
+    assert_eq!(st.get_total_balance(acct_id), expected_final);
+
+    use incrementalmerkletree::Position;
+    let expected_checkpoints_p0: Vec<(BlockHeight, ShieldedProtocol, Option<Position>)> = [
+        (99999, None),
+        (100000, Some(0)),
+        (100001, Some(1)),
+        (100002, Some(1)),
+        (100006, Some(1)),
+        (100013, Some(3)),
+        (100014, Some(5)),
+        (100020, Some(6)),
+    ]
+    .into_iter()
+    .map(|(h, pos)| {
+        (
+            BlockHeight::from(h),
+            P0::SHIELDED_PROTOCOL,
+            pos.map(Position::from),
+        )
+    })
+    .collect();
+
+    let expected_checkpoints_p1: Vec<(BlockHeight, ShieldedProtocol, Option<Position>)> = [
+        (99999, None),
+        (100000, None),
+        (100001, None),
+        (100002, Some(0)),
+        (100006, Some(0)),
+        (100013, Some(0)),
+        (100014, Some(2)),
+        (100020, Some(2)),
+    ]
+    .into_iter()
+    .map(|(h, pos)| {
+        (
+            BlockHeight::from(h),
+            P1::SHIELDED_PROTOCOL,
+            pos.map(Position::from),
+        )
+    })
+    .collect();
+
+    let actual_checkpoints = st.get_checkpoint_history().unwrap();
+
+    assert_eq!(
+        actual_checkpoints
+            .iter()
+            .filter(|(_, p, _)| p == &P0::SHIELDED_PROTOCOL)
+            .cloned()
+            .collect::<Vec<_>>(),
+        expected_checkpoints_p0
+    );
+    assert_eq!(
+        actual_checkpoints
+            .iter()
+            .filter(|(_, p, _)| p == &P1::SHIELDED_PROTOCOL)
+            .cloned()
+            .collect::<Vec<_>>(),
+        expected_checkpoints_p1
+    );
+}
+
+#[cfg(feature = "orchard")]
+pub(crate) fn multi_pool_checkpoints_with_pruning<
+    P0: ShieldedPoolTester,
+    P1: ShieldedPoolTester,
+>() {
+    let mut st = TestBuilder::new()
+        .with_block_cache()
+        .with_account_from_sapling_activation(BlockHash([0; 32])) // TODO: Allow for Orchard
+        // activation after Sapling
+        .build();
+
+    let account = st.test_account().cloned().unwrap();
+
+    let p0_fvk = P0::random_fvk(&mut st.rng);
+    let p1_fvk = P1::random_fvk(&mut st.rng);
+
+    let note_value = NonNegativeAmount::const_from_u64(10000);
+    // Generate 100 P0 blocks, then 100 P1 blocks, then another 100 P0 blocks.
```

Comment:
Bug: this documentation does not match what the code is actually doing (which is much more tightly interleaving P0 and P1 blocks).

Reply:
This is the bit where I am _still_ unable to actually reproduce the error except in a mainnet wallet, and I'm kind of at wit's end.